api/v1alpha1/templates_common.go

```diff
@@ -73,3 +76,66 @@ type TemplateValidationStatus struct {
 	// Valid indicates whether the template passed validation or not.
 	Valid bool `json:"valid"`
 }
+
+type providersType int
+
+const (
+	bootstrapProvidersType providersType = iota
+	controlPlaneProvidersType
+	infrastructureProvidersType
+)
+
+func parseProviders[T any](providersGetter interface{ GetSpecProviders() ProvidersTupled }, typ providersType, annotations map[string]string, validationFn func(string) (T, error)) ([]ProviderTuple, error) {
+	pspec, anno := getProvidersSpecAnno(providersGetter, typ)
+	if len(pspec) > 0 {
+		return pspec, nil
+	}
+
+	providers := annotations[anno]
+	if len(providers) == 0 {
+		return []ProviderTuple{}, nil
+	}
+
+	var (
+		splitted = strings.Split(providers, ",")
+		pstatus  = make([]ProviderTuple, 0, len(splitted))
+		merr     error
+	)
+	for _, v := range splitted {
+		v = strings.TrimSpace(v)
+		nVerOrC := strings.SplitN(v, " ", 2)
```

Comment:
let's discuss the format separately

Reply:
reconciled on the `;` to be used as the separator, and I will propagate it to the providers' list without constraints/exact versions, which is now only applicable for the `servicetemplates` to avoid inconsistency